synapse/handlers/room_hierarchy.py

```diff
@@ -0,0 +1,274 @@
+
+
+logger = logging.getLogger(__name__)
+
+
+class RoomHierarchyHandler:
+    """Provides methods for walking over space hierarchies.
+
+    Also see `RoomSummaryHandler`, which has similar functionality.
+    """
+
+    def __init__(self, hs: "HomeServer"):
+        self._store = hs.get_datastore()
+        self._federation_client = hs.get_federation_client()
+
+        self._server_name = hs.hostname
+
+    async def get_room_descendants(
+        self, space_id: str, via: Optional[Iterable[str]] = None
+    ) -> Tuple[Sequence[Tuple[str, Iterable[str]]], Sequence[str]]:
+        """Gets the children of a space, recursively.
+
+        Args:
+            space_id: The room ID of the space.
+
+        Returns:
+            A tuple containing:
+             * A list of (room ID, via) tuples, representing the descendants of the
+               space. `space_id` is included in the list.
+             * A list of room IDs whose children could not be fully listed.
+               Rooms in this list are either spaces not known locally, and thus require
+               listing over federation, or are unknown rooms or subspaces completely
+               inaccessible to the local homeserver which may contain further rooms.
+
+               This list is a subset of the previous list, except it may include
+               `space_id`.
+        """
+        via = via or []
+
+        # (room ID, via, federation room chunks)
+        todo: List[Tuple[str, Iterable[str], Mapping[str, Optional[JsonDict]]]] = [
+            (space_id, via, {})
+        ]
+        descendants: List[Tuple[str, Iterable[str]]] = []
+
+        seen = {space_id}
+
+        inaccessible_room_ids: List[str] = []
+
+        while todo:
+            space_id, via, federation_room_chunks = todo.pop()
+            descendants.append((space_id, via))
+            try:
+                (
+                    is_in_room,
+                    children,
+                    federation_room_chunks,
+                ) = await self._get_room_children(space_id, via, federation_room_chunks)
+            except SynapseError:
+                # Could not list children over federation
+                inaccessible_room_ids.append(space_id)
+                continue
+
+            for child_room_id, child_via in reversed(children):
+                if child_room_id in seen:
+                    continue
+
+                seen.add(child_room_id)
+
+                # Queue up the child for processing.
+                # The child may not actually be a space, but that's checked by
+                # `_get_room_children`.
+                todo.append((child_room_id, child_via, federation_room_chunks))
+
+            # Children were retrieved over federation, which is not guaranteed to be
+            # the full list.
+            if not is_in_room:
+                inaccessible_room_ids.append(space_id)
+
+        return descendants, inaccessible_room_ids
+
+    async def _get_room_children(
```

Comment:
`get_rooms_in_space` doesn't convey the difference between direct children and descendants that I want.

`get_space_(children|descendants)` sounds plausible. I'm actually confused about the room vs space terminology because in `RoomSummaryHandler`, we have space summaries and room hierarchies.

Reply:
> `get_rooms_in_space` doesn't convey the difference between direct children and descendants that I want.

Hm, I guess that's fair enough. I don't think I'd imagine anything other than direct children for that, but it's hard to say I can't see where others would be coming from if they did. Best avoid ambiguous terminology after all..

> get_space_(children|descendants) sounds plausible. I'm actually confused about the room vs space terminology because in RoomSummaryHandler, we have space summaries and room hierarchies.

This is probably a typical case of naming is hard and everything is awful :/.
However you **are** checking that the room is a space (bailing if it's not), so personally speaking I would call it `get_space_children` (especially since `m.space.child` exists).

Reply:
Room hierarchies might be because the leaves in the tree are not spaces, they are merely rooms. Of course, the root node and interior nodes, being spaces, are also rooms, so the 'Room' in 'Room hierarchies' is so because it's the most general. I'm just making this up so might have nothing to do with the real reason...

Reply:
I renamed a bunch of things in this PR to refer to spaces. Calling them spaces rather than rooms is a lot more intuitive for me at least: I had trouble finding `room_summary.py` because I was looking for something called `space_*` first.

Reply:
Yeah, happy to conclude that it might be the most intuitive name.